tools/iana_to_rflx.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env -S python3 -O
+
+import argparse
+import re
+import string
+import sys
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List, Optional, TextIO, Tuple, Union
+from xml.etree import ElementTree
+from xml.etree.ElementTree import Element
+
+import rflx.specification.const
+
+NAMESPACE = {"iana": "http://www.iana.org/assignments"}
+RESERVED_WORDS = "|".join(rflx.specification.const.RESERVED_WORDS)
+DUPLICATES: List[str] = []
+
+
+def iana_to_rflx(xml_input: TextIO, always_valid: bool, output_file: Optional[Path] = None) -> None:
+    xml_str = xml_input.read()
+    root = ElementTree.fromstring(xml_str)
+    registry_id = root.get("id")
+    if registry_id is None:
+        raise IANAError("No registry ID found")
+    package_name = _normalize_name(registry_id)
+    last_updated = root.find("iana:updated", NAMESPACE)
+    title = root.find("iana:title", NAMESPACE)
+
+    with open(
+        f"{package_name.lower() + '.rflx' if output_file is None else output_file}", "w+"
+    ) as file:
+        file.write(f"-- AUTOMATICALLY GENERATED BY {__file__}. DO NOT EDIT.\n")
+        file.write(f"-- Generation date: {datetime.now().strftime('%Y-%m-%d')}\n")
+        if title is not None:
+            file.write(f"-- {title.text}\n")
+        if last_updated is not None:
+            file.write(f"-- Registry last updated on {last_updated.text}\n\n")
+        file.write(f"package {package_name} is\n\n")
+        for registry in root.findall(root.tag):
+            write_registry(registry, always_valid, file)
+        file.write(f"end {package_name};")
+
+
+def write_registry(
+    registry: Element,
+    always_valid: bool,
+    file: TextIO,
+) -> None:
+    if registry.find("iana:record", NAMESPACE) is None:
+        return
+    records = registry.findall("iana:record", NAMESPACE)
+    title = registry.find("iana:title", NAMESPACE)
+    if title is None or title.text is None:
+        raise IANAError("could not find registry title")
+    registry_title = _normalize_name(title.text)
+    normalized_records = _normalize_records(records, registry_title)
+    if not normalized_records:
+        return
+
+    file.write(f"{'':<3}type {registry_title} is\n")
+    file.write(f"{'':<6}(\n")
+    size = max((record.bit_length for record in normalized_records))
```

Comment:
I think such types do not always need to have a size that is a multiple of 8. Did you find concrete cases, where the size cannot be inferred by the highest value?

Reply:
No, but when the type is used in another specification, it could be required to have a size that is byte-aligned. I guess it would be easiest to edit the generated file manually in such a case.

Reply:
IMHO it shouldn't be necessary to edit the generated files manually, so that a specification can be regenerated using an updated IANA document without the need for maintaining manual changes. But as long as there is no concrete use case, it doesn't make sense to add any specific functionality.